File: Pastura/PasturaTests/Engine/ScenarioSemanticLinterTests.swift
// One @Test per lint rule (added by later plan items) keeps the
// finding-mode mapping legible. This file starts with the core-type
// contract (skeleton) — rule-specific tests land as R1–R17 are filled in.
import Testing

@testable import Pastura

@Suite(.timeLimit(.minutes(1)))
struct ScenarioSemanticLinterTests {
  let linter = ScenarioSemanticLinter()

  // MARK: - Core contract

  @Test func minimalValidScenarioHasNoFindings() {
    let scenario = makeScenario(agents: 2, rounds: 1, phases: [Phase(type: .speakAll)])
    #expect(linter.lint(scenario).isEmpty)
  }

  @Test func findingEqualityHoldsForIdenticalFields() {
    let lhs = LintFinding(
      ruleID: "example-rule", severity: .warning, message: "msg", phaseIndex: 2)
    let rhs = LintFinding(
      ruleID: "example-rule", severity: .warning, message: "msg", phaseIndex: 2)
    #expect(lhs == rhs)
  }

  @Test func findingEqualityDistinguishesEachField() {
    let base = LintFinding(
      ruleID: "example-rule", severity: .warning, message: "msg", phaseIndex: 2)
    #expect(
      base
        != LintFinding(
          ruleID: "other-rule", severity: .warning, message: "msg", phaseIndex: 2))
    #expect(
      base
        != LintFinding(
          ruleID: "example-rule", severity: .error, message: "msg", phaseIndex: 2))
    #expect(
      base
        != LintFinding(
          ruleID: "example-rule", severity: .warning, message: "other", phaseIndex: 2))
    #expect(
      base
        != LintFinding(
          ruleID: "example-rule", severity: .warning, message: "msg", phaseIndex: nil))
  }

  @Test func scenarioLevelFindingUsesNilPhaseIndex() {
    let finding = LintFinding(
      ruleID: "scenario-rule", severity: .info, message: "msg", phaseIndex: nil)
    #expect(finding.phaseIndex == nil)
  }

  // MARK: - Severity

  @Test func severityCasesAreDistinct() {
    #expect(LintSeverity.error != LintSeverity.warning)
    #expect(LintSeverity.warning != LintSeverity.info)
    #expect(LintSeverity.error != LintSeverity.info)
  }

  @Test func severityOrdersErrorAboveWarningAboveInfo() {
    // Comparable is by blocking-strength: error is the most severe.
    #expect(LintSeverity.info < LintSeverity.warning)
    #expect(LintSeverity.warning < LintSeverity.error)
    #expect(LintSeverity.info < LintSeverity.error)
  }

  // MARK: - Helpers

  // Internal (not private) so later sibling-file extensions can build the
  // many phase-list permutations rule tests need from one factory.
  func makeScenario(agents: Int, rounds: Int, phases: [Phase]) -> Scenario {
    Scenario(
      id: "test", name: "Test", description: "Test",
      language: "ja",
      agentCount: agents, rounds: rounds, context: "Context",
      personas: (0..<agents).map { Persona(name: "A\($0)", description: "D") },
      phases: phases
    )
  }
}

File: shared/engine/src/commonTest/kotlin/com/pastura/engine/ScenarioSemanticLinterTests.kt
package com.pastura.engine

import com.pastura.models.Phase
import com.pastura.models.PhaseType
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Core-contract tests for [ScenarioSemanticLinter] / [LintFinding] /
 * [LintSeverity] — mirrors
 * `Pastura/PasturaTests/Engine/ScenarioSemanticLinterTests.swift` 1:1 (its
 * "Core contract" and "Severity" sections; the Ordering-rule tests land with
 * D2a item 2, which implements `orderingFindings`). Every function name below
 * matches its Swift twin exactly — see that file if a name here looks odd.
 */
class ScenarioSemanticLinterTests {

    private val linter = ScenarioSemanticLinter()

    // MARK: - Core contract

    @Test
    fun minimalValidScenarioHasNoFindings() {
        val scenario = makeLinterScenario(
            agents = 2, rounds = 1, phases = listOf(Phase(type = PhaseType.SPEAK_ALL)),
        )
        assertTrue(linter.lint(scenario).isEmpty())
    }

    @Test
    fun findingEqualityHoldsForIdenticalFields() {
        val lhs = LintFinding(
            ruleId = "example-rule", severity = LintSeverity.WARNING, message = "msg", phaseIndex = 2,
        )
        val rhs = LintFinding(
            ruleId = "example-rule", severity = LintSeverity.WARNING, message = "msg", phaseIndex = 2,
        )
        assertEquals(lhs, rhs)
    }

    @Test
    fun findingEqualityDistinguishesEachField() {
        val base = LintFinding(
            ruleId = "example-rule", severity = LintSeverity.WARNING, message = "msg", phaseIndex = 2,
        )
        assertNotEquals(
            base,
            LintFinding(
                ruleId = "other-rule", severity = LintSeverity.WARNING, message = "msg", phaseIndex = 2,
            ),
        )
        assertNotEquals(
            base,
            LintFinding(
                ruleId = "example-rule", severity = LintSeverity.ERROR, message = "msg", phaseIndex = 2,
            ),
        )
        assertNotEquals(
            base,
            LintFinding(
                ruleId = "example-rule", severity = LintSeverity.WARNING, message = "other", phaseIndex = 2,
            ),
        )
        assertNotEquals(
            base,
            LintFinding(
                ruleId = "example-rule", severity = LintSeverity.WARNING, message = "msg", phaseIndex = null,
            ),
        )
    }

    @Test
    fun scenarioLevelFindingUsesNilPhaseIndex() {
        val finding = LintFinding(
            ruleId = "scenario-rule", severity = LintSeverity.INFO, message = "msg", phaseIndex = null,
        )
        assertNull(finding.phaseIndex)
    }

    // MARK: - Severity

    @Test
    fun severityCasesAreDistinct() {
        assertNotEquals(LintSeverity.ERROR, LintSeverity.WARNING)
        assertNotEquals(LintSeverity.WARNING, LintSeverity.INFO)
        assertNotEquals(LintSeverity.ERROR, LintSeverity.INFO)
    }

    @Test
    fun severityOrdersErrorAboveWarningAboveInfo() {
        // Comparable is by blocking-strength: error is the most severe. Kotlin's
        // enum-declaration-order Comparable matches Swift's explicit Comparable
        // conformance here because LintSeverity is declared INFO, WARNING, ERROR.
        assertTrue(LintSeverity.INFO < LintSeverity.WARNING)
        assertTrue(LintSeverity.WARNING < LintSeverity.ERROR)
        assertTrue(LintSeverity.INFO < LintSeverity.ERROR)
    }
}
